src/ripple/basics/Expected.h

```diff
@@ -0,0 +1,244 @@
+    {
+    }
+
+    constexpr explicit Unexpected(E&& e) : val_(std::move(e))
+    {
+    }
+
+    constexpr const E&
+    value() const&
+    {
+        return val_;
+    }
+
+    constexpr E&
+    value() &
+    {
+        return val_;
+    }
+
+    constexpr E&&
+    value() &&
+    {
+        return std::move(val_);
+    }
+
+    constexpr const E&&
+    value() const&&
+    {
+        return std::move(val_);
+    }
+
+private:
+    E val_;
+};
+
+// Unexpected deduction guide that converts array to const*.
+template <typename E, std::size_t N>
+Unexpected(E (&)[N]) -> Unexpected<E const*>;
+
+// Definition of Expected.  All of the machinery comes from boost::result.
+template <class T, class E>
+class [[nodiscard]] Expected
+    : private boost::outcome_v2::result<T, E, detail::throw_policy>
+{
+    using Base = boost::outcome_v2::result<T, E, detail::throw_policy>;
+
+public:
+    template <
+        typename U,
+        typename = std::enable_if_t<std::is_convertible_v<U, T>>>
+    constexpr Expected(U r) : Base(T{std::forward<U>(r)})
+    {
+    }
+
+    template <
+        typename U,
+        typename = std::enable_if_t<std::is_convertible_v<U, E>>>
+    constexpr Expected(Unexpected<U> e) : Base(E{std::forward<U>(e.value())})
+    {
+    }
+
+    constexpr bool has_value() const
+    {
+        return Base::has_value();
+    }
+
+    constexpr T const& value() const
+    {
+        return Base::value();
+    }
+
+    constexpr T& value()
+    {
+        return Base::value();
+    }
+
+    constexpr E const& error() const
+    {
+        return Base::error();
+    }
+
+    constexpr E& error()
+    {
+        return Base::error();
+    }
+
+    constexpr explicit operator bool() const
+    {
+        return has_value();
+    }
+
+    // Add operator* and operator-> so the Expected API looks a bit more like
+    // what std::expected is likely to look like.  See:
+    // http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p0323r10.html
+    [[nodiscard]] constexpr T& operator*()
+    {
+        return this->value();
+    }
+
+    [[nodiscard]] constexpr T const& operator*() const
+    {
+        return this->value();
+    }
+
+    [[nodiscard]] constexpr T* operator->()
+    {
+        return &this->value();
+    }
+
+    [[nodiscard]] constexpr T const* operator->() const
+    {
+        return &this->value();
+    }
+};
+
+// Specialization of Expected<void, E>.  Allows returning either success
+// (without a value) or the reason for the failure.
+template <class E>
+class [[nodiscard]] Expected<void, E>
+    : private boost::outcome_v2::result<void, E, detail::throw_policy>
+{
+    using Base = boost::outcome_v2::result<void, E, detail::throw_policy>;
+
+public:
+    // We'd like to avoid a default constructor.  That means someone needs
+    // explicitly say they intent to return success.  Seems like
+    // std::true_type is a good way to do that...
+    constexpr Expected(std::true_type) : Base(boost::outcome_v2::success())
```

Comment:
It looks like the proposed standard `expected` has a default constructor for a `void` return (ref: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p0323r9.html#expected.object.ctor). I think a default constructor makes sense in this case, and give the standard has it, I'd like to remove this constructor and replace it with a default constructor.

Reply:
Thanks for catching the default constructor for `Expected<void, E>`.  I looked for something like `boost::outcome_v2::success()` in the paper and didn't find it, so I punted.  Matching the currently proposed version of the paper seems like the best approach.  I've included an additional `[FOLD]` to make this change.